docs/cookbook/images-gridfs.rst

```diff
@@ -0,0 +1,643 @@
+
+.. note::
+
+    The route ``product_cache_image`` is defined via an annotation on the ``ImagineController::imageAction`` class method below this example.
+
+.. code-block:: php
+
+    namespace AppBundle\Imagine\Cache\Resolver;
+
+    use AppBundle\Document\Product\Image\Cache;
+    use Doctrine\ODM\MongoDB\DocumentManager;
+    use Doctrine\ODM\MongoDB\DocumentRepository;
+    use Liip\ImagineBundle\Binary\BinaryInterface;
+    use Liip\ImagineBundle\Imagine\Cache\Resolver\ResolverInterface;
+    use Symfony\Component\Routing\RouterInterface;
+
+    class GridFSResolver implements ResolverInterface
+    {
+        /**
+         * @var DocumentManager
+         */
+        private $documentManager;
+
+        /**
+         * @var string
+         */
+        private $class;
+
+        /**
+         * @var RouterInterface
+         */
+        private $router;
+
+        /**
+         * @param DocumentManager $documentManager
+         * @param string $class
+         * @param RouterInterface $router
+         */
+        public function __construct(DocumentManager $documentManager, string $class, RouterInterface $router)
+        {
+            $this->documentManager = $documentManager;
+            $this->class = $class;
+            $this->router = $router;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function isStored($path, $filter)
+        {
+            return $this->findCacheFile($path, $filter) !== null;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function resolve($path, $filter)
+        {
+            $cache = $this->findCacheFile($path, $filter);
+
+            return $this->router->generate('product_cache_image', ['id' => $cache->getId()], RouterInterface::ABSOLUTE_URL);
+        }
+
+        /**
+         * @param string $path
+         * @param string $filter
+         *
+         * @return Cache|null
+         */
+        private function findCacheFile($path, string $filter)
+        {
+            return $this->getRepository()->findOneBy(['filename' => $path, 'metadata.filter' => $filter]);
+        }
+
+        /**
+         * {@inheritdoc}
+         *
+         * @throws GridFSException
+         */
+        public function store(BinaryInterface $binary, $path, $filter)
+        {
+            $file = tempnam(sys_get_temp_dir(), 'GridFSResolver');
+
+            if (file_put_contents($file, $binary->getContent()) === false) {
+                // We're using a custom exception to make it explicit catchable
+                throw new GridFSException("Could not write cache file '{$file}' to disk");
+            }
+
+            try {
+                $cache = new Cache($file, $binary->getMimeType(), $path, $filter);
+
+                $this->documentManager->persist($cache);
+                $this->documentManager->flush();
+            } finally {
```

Comment:
Catching the exception is omitted here as it's nothing the `GridFSResolver` is able to handle. The `try`-`finally` is only used to ensure the temporary file is removed from disk (regardless if the `flush()` is successful or not.